### PR TITLE
Simplify ProjectionPushdown and make it more general

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Follow on to https://github.com/apache/arrow-datafusion/pull/8073

## Rationale for this change

I noticed some ways the code could be simplified and made more general while reviewing the PR, 

## What changes are included in this PR?

1. Use `TreeNode::transform_down` instead of a manual tree walk to rewrite expressions
2. pull a common check out of each branch

## Are these changes tested?

Yes, by existing tests

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/projection_pushdown.rs [hunk 1]
```diff
@@ -100,6 +97,13 @@ pub fn remove_unnecessary_projections(
         if is_projection_removable(projection) {
             return Ok(Transformed::Yes(projection.input().clone()));
         }
+
+        // If the projection does not narrow the the schema, we should not try
+        // to push it down
+        if projection.expr().len() >= projection.input().schema().fields().len() {
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/projection_pushdown.rs:103]: I pulled this out of the individual `try_swapping_*` calls

File: datafusion/core/src/physical_optimizer/projection_pushdown.rs [hunk 2]
```diff
@@ -791,119 +767,42 @@ fn update_expr(
...
-        })
-        .collect::<Vec<_>>();
-
-    if new_when_then.len() != case.when_then_expr().len()
-        || case.expr().is_some() && new_case.is_none()
-        || case.else_expr().is_some() && new_else.is_none()
-    {
-        return Ok(None);
-    }
+    let new_expr = expr
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/projection_pushdown.rs:772]: I rewrote this logic to use TreeNode recursion / rewriting rather  than a manual recursion, which I think is both less code and is more general (handles all PhysicalExprs, not just the ones explicitly checked for here)

**Comment by alamb** [comment]
Some test is failing -- I will debug it later

File: datafusion/common/src/tree_node.rs [hunk 3]
```diff
@@ -149,6 +149,19 @@ pub trait TreeNode: Sized {
         Ok(new_node)
     }
 
+    /// Convenience utils for writing optimizers rule: recursively apply the given 'op' first to all of its
+    /// children and then itself(Postorder Traversal) using a mutable function, `F`.
+    /// When the `op` does not apply to a given node, it is left unchanged.
+    fn transform_up_mut<F>(self, op: &mut F) -> Result<Self>
```
Comment (alamb) [inline datafusion/common/src/tree_node.rs:155]: `mut` variant is needed to support changing a variable in the closure (updating `state`)

**Comment by alamb** [comment]
cc @berkaysynnada

File: datafusion/core/src/physical_optimizer/projection_pushdown.rs [hunk 4]
```diff
@@ -791,119 +788,65 @@ fn update_expr(
...
+                }
+            }
+        });
+
+    new_expr.map(|new_expr| {
+        if state == RewriteState::RewrittenValid {
+            Some(new_expr)
+        } else {
+            None
+        }
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/projection_pushdown.rs:848]:
```suggestion
    new_expr.map(|e| (state == RewriteState::RewrittenValid).then_some(e))
```

**Review by ozankabak — COMMENTED** [review]
LGTM apart from one idiom suggestion. I'll let @berkaysynnada take a look and make sure all is right, then this will be good to go from our perspective. Thanks Andrew

File: datafusion/core/src/physical_optimizer/projection_pushdown.rs [hunk 5]
```diff
@@ -791,119 +788,65 @@ fn update_expr(
...
+                    },
+                );
+                if let Some(new_col) = new_col {
+                    state = RewriteState::RewrittenValid;
+                    Ok(Transformed::Yes(new_col))
+                } else {
+                    // didn't find a rewrite, stop trying
+                    state = RewriteState::RewrittenInvalid;
+                    Ok(Transformed::No(expr))
+                }
```
Comment (berkaysynnada) [inline datafusion/core/src/physical_optimizer/projection_pushdown.rs:839]:
```suggestion
 state = RewriteState::RewrittenInvalid;
                projected_exprs
                    .iter()
                    .enumerate()
                    .find_map(|(index, (projected_expr, alias))| {
                        projected_expr.as_any().downcast_ref::<Column>().and_then(
                            |projected_column| {
                                column.name().eq(projected_column.name()).then(|| {
                                    state = RewriteState::RewrittenValid;
                                    Arc::new(Column::new(alias, index)) as _
                                })
                            },
                        )
                    })
                    .map_or(Ok(Transformed::No(expr)), |new_column| {
                        Ok(Transformed::Yes(new_column))
                    })
```

**Comment by berkaysynnada** [comment]
Thanks @alamb, that function will possibly be used for future iterations as well. It's better now.

**Review by ozankabak — APPROVED** [review]
LGTM, thank you

**Pull request merged**